File: Programmers/Level1/220824/Java/MockTest.java
// 1번 수포자 : 1, 2, 3, 4, 5, 1, 2, 3, 4, 5, ...
// 2번 수포자 : 2, 1, 2, 3, 2, 4, 2, 5, 2, 1, 2, 3, 2, 4, 2, 5 ...
// 3번 수포자 : 3, 3, 1, 1, 2, 2, 4, 4, 5, 5, 3, 3, 1, 1 ...

public class MockTest {
  public int[] solution(int[] answers) {
    int[] answer = {};

    int[] firstAnswers = new int[]{1, 2, 3, 4, 5}; // 5
    int[] secondAnswers = new int[]{2, 1, 2, 3, 2, 4, 2, 5}; // 8
    int[] thirdAnswers = new int[]{3, 3, 1, 1, 2, 2, 4, 4, 5, 5}; // 10

    int firstCount = gradingFirstAnswers(answers, firstAnswers);
    int secondCount = gradingSecondAnswers(answers, secondAnswers);
    int thirdCount = gradingThirdAnswers(answers, thirdAnswers);


    if (firstCount > secondCount && firstCount > thirdCount) {
      return new int[]{1};
    }
    if (secondCount > firstCount && secondCount > thirdCount) {
      return new int[]{2};
    }
    if (thirdCount > secondCount && thirdCount > firstCount) {
      return new int[]{3};
    }

    if (firstCount == secondCount && firstCount > thirdCount) {
      return new int[]{1, 2};
    }

    if (firstCount == thirdCount && firstCount > secondCount) {
      return new int[]{1, 3};
    }

    if (secondCount == thirdCount && secondCount > firstCount) {
      return new int[]{2, 3};
    }

    if (firstCount == secondCount && firstCount == thirdCount) {
      return new int[]{1, 2, 3};
    }

    return answer;
  }


  public int gradingFirstAnswers(int[] answers, int[] firstAnswers) {
    int firstCount = 0;
    for (int i = 0; i < answers.length; i += 1) {
      if (firstAnswers[i % 5] == answers[i]) {
        firstCount += 1;
      }
    }
    return firstCount;
  }

  public int gradingSecondAnswers(int[] answers, int[] secondAnswers) {
    int secoundCount = 0;
    for (int i = 0; i < answers.length; i += 1) {
      if (secondAnswers[i % 8] == answers[i]) {
        secoundCount += 1;
      }
    }
    return secoundCount;
  }

  public int gradingThirdAnswers(int[] answers, int[] thirdAnswers) {
    int thirdCount = 0;
    for (int i = 0; i < answers.length; i += 1) {
      if (thirdAnswers[i % 10] == answers[i]) {
        thirdCount += 1;
      }
    }
    return thirdCount;
  }
}

File: Programmers/Level1/220825/JavaScript/MockTest.js
export function solution(answers) {
  const answer = [];

  const firstCount = {
    count: 0,
  };

  const secoundCount = {
    count: 0,
  };

  const thirdCount = {
    count: 0,
  };

  firstCount.count = firstCountCompute(firstCount.count, answers);
  secoundCount.count = secoundCountCompute(secoundCount.count, answers);
  thirdCount.count = thirdCountCompute(thirdCount.count, answers);

  const maximumCount = Math.max(firstCount.count, secoundCount.count, thirdCount.count);

  if (firstCount.count === maximumCount) {
    answer.push(1);
  }

  if (secoundCount.count === maximumCount) {
    answer.push(2);
  }

  if (thirdCount.count === maximumCount) {
    answer.push(3);
  }

  return answer;
}

export function firstCountCompute(count, answers) {
  const firstAnswers = [1, 2, 3, 4, 5];

  for (let i = 0; i < answers.length; i += 1) {
    if (firstAnswers[i % 5] === answers[i]) {
      count += 1;
    }
  }
  return count;
}

export function secoundCountCompute(count, answers) {
  const secondAnswers = [2, 1, 2, 3, 2, 4, 2, 5];

  for (let i = 0; i < answers.length; i += 1) {
    if (secondAnswers[i % 8] === answers[i]) {
      count += 1;
    }
  }
  return count;
}

export function thirdCountCompute(count, answers) {
  const thirdAnswers = [3, 3, 1, 1, 2, 2, 4, 4, 5, 5];
  for (let i = 0; i < answers.length; i += 1) {
    if (thirdAnswers[i % 10] === answers[i]) {
      count += 1;
    }
  }
  return count;
}
